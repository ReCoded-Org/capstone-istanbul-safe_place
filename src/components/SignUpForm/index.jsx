import React from "react";
import { useForm } from "react-hook-form";
import Joi from "joi";
import { joiResolver } from "@hookform/resolvers/joi";
import { useTranslation } from "react-i18next";
import InputErrorMessage from "../InputErrorMessage";
import { getErrorClass } from "../../utils/formHelpers";
import { EMAIL_ALREADY_TAKEN_ERROR } from "../../utils/authHelpers";
import "./index.scss";

const schema = Joi.object({
  firstName: Joi.string().required().messages({
    "string.empty": `Please provide your first name`,
  }),
  lastName: Joi.string().required().messages({
    "string.empty": `Please provide your last name`,
  }),
  email: Joi.string().required().email({ tlds: {} }).messages({
    "string.empty": `Please provide your email`,
    "string.email": `Please provide a valid email`,
  }),
  password: Joi.string().required().min(8).max(64).strict().messages({
    "string.empty": `Please provide a password`,
  }),
  confirmPassword: Joi.string()
    .valid(Joi.ref("password"))
    .required()
    .strict()
    .messages({
      "any.only": "Those passwords didn't match",
    }),
  acceptTerms: Joi.valid("accepted").required().messages({
    "any.only": `You must accept our terms of service and privacy policy`,
  }),
});

export default function SignUpForm({ submit }) {
  const { register, handleSubmit, errors, reset, setError } = useForm({
    resolver: joiResolver(schema),
  });
  const { t } = useTranslation();

  const onSubmit = async (data) => {
    const status = await submit(data);
    switch (status) {
      case "auth/email-already-in-use":
        setError("email", {
          type: EMAIL_ALREADY_TAKEN_ERROR,
          message:
            "Email is already taken! Please enter another one or sign in.",
        });
        break;
      case "succeed":
        reset();
        break;
      default:
        reset();
    }
  };

  return (
    <div className="signUpForm">
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className="formInputGroup">
          <div className="formInput">
            <input
              className={getErrorClass(errors.firstName)}
              type="text"
              placeholder={t("signUp.firstName")}
              name="firstName"
              aria-label="firstName"
              ref={register()}
            />
            <InputErrorMessage error={errors.firstName} />
          </div>

          <div className="formInput">
            <input
              className={getErrorClass(errors.lastName)}
              type="text"
              placeholder={t("signUp.lastName")}
              name="lastName"
              aria-label="lastName"
              ref={register()}
            />
            <InputErrorMessage error={errors.lastName} />
          </div>
        </div>

        <div className="formInput">
          <input
            className={getErrorClass(errors.email)}
            type="email"
            placeholder={t("signUp.emailAddress")}
            name="email"
            aria-label="email"
            ref={register()}
          />
          <InputErrorMessage error={errors.email} />
        </div>

        <div className="formInputGroup">
          <div className="formInput">
            <input
              className={getErrorClass(errors.password)}
              type="password"
              placeholder={t("signUp.password")}
              name="password"
              aria-label="password"
              ref={register()}
            />
            <InputErrorMessage error={errors.password} />
          </div>

          <div className="formInput">
            <input
              className={getErrorClass(errors.confirmPassword)}
              type="password"
              placeholder={t("signUp.confirmPassword")}
              name="confirmPassword"
              aria-label="confirmPassword"
              ref={register()}
            />
            <InputErrorMessage error={errors.confirmPassword} />
          </div>
        </div>

        <div className="formInput">
          <div className="termsConditions">
            <div>
              <input
                className={getErrorClass(errors.acceptTerms)}
                type="checkbox"
                id="acceptTerms"
                name="acceptTerms"
                value="accepted"
                aria-label="acceptTerms"
                ref={register()}
              />
            </div>
            <div>
              <label
                htmlFor="acceptTerms"
                dangerouslySetInnerHTML={{
                  __html: t("signUp.acceptTerms", {
                    terms: `<a href="/terms">${t("signUp.terms")}</a>`,
                    privacy: `<a href="/privacy">Privacy Policy</a>`,
                  }),
                }}
              />
            </div>
          </div>
          <InputErrorMessage error={errors.acceptTerms} />
        </div>

        <button type="submit" className="submitBtn">
          {t("signUp.signUp")}
        </button>

        <p
          dangerouslySetInnerHTML={{
            __html: t("signUp.haveAccount", {
              signIn: `<a href="/signin">${t("signUp.signIn")}</a>`,
            }),
          }}
        />
      </form>
    </div>
  );
}
