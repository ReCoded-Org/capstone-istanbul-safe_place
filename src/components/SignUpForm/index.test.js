import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import SignUpForm from "./index";

describe("SignUpForm", () => {
  it("renders correctly and matches the snapshot", () => {
    const tree = renderer.create(<SignUpForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  const mockSubmit = jest.fn(
    (firstName, lastName, email, password, confirmPassword, acceptTerms) => {
      return Promise.resolve({
        firstName,
        lastName,
        email,
        password,
        confirmPassword,
        acceptTerms,
      });
    }
  );

  beforeEach(() => {
    render(<SignUpForm submit={mockSubmit} />);
  });

  it("should display required error when value is empty", async () => {
    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(5);
    expect(mockSubmit).not.toBeCalled();
  });

  it("should display matching error when email is invalid", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
      target: {
        value: "name",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
      target: {
        value: "surname",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.click(screen.getByRole("checkbox"));

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(mockSubmit).not.toBeCalled();
    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe(
      "name"
    );
    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe(
      "surname"
    );
    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe("test");
    expect(screen.getByPlaceholderText(/^password/i).value).toBe("password");
    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
      "password"
    );
    expect(screen.getByRole("checkbox").checked).toBe(true);
  });

  it("should display matching error when email domain is invalid", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
      target: {
        value: "name",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
      target: {
        value: "surname",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@test.co",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.click(screen.getByRole("checkbox"));

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(mockSubmit).not.toBeCalled();
    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe(
      "name"
    );
    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe(
      "surname"
    );
    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe(
      "test@test.co"
    );
    expect(screen.getByPlaceholderText(/^password/i).value).toBe("password");
    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
      "password"
    );
    expect(screen.getByRole("checkbox").checked).toBe(true);
  });

  it("should display min length error when password is short", async () => {
    const MIN_CHAR = 8;
    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
      target: {
        value: "name",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
      target: {
        value: "surname",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
      target: {
        value: `${"a".repeat(MIN_CHAR - 1)}`,
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
      target: {
        value: `${"a".repeat(MIN_CHAR - 1)}`,
      },
    });

    fireEvent.click(screen.getByRole("checkbox"));

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(mockSubmit).not.toBeCalled();
    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe(
      "name"
    );
    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe(
      "surname"
    );
    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe(
      "test@mail.com"
    );
    expect(screen.getByPlaceholderText(/^password/i).value).toBe(
      `${"a".repeat(MIN_CHAR - 1)}`
    );
    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
      `${"a".repeat(MIN_CHAR - 1)}`
    );
    expect(screen.getByRole("checkbox").checked).toBe(true);
  });

  it("should display max length error when password is long", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
      target: {
        value: "name",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
      target: {
        value: "surname",
      },
    });

    const MAX_CHAR = 64;
    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
      target: {
        value: `${"a".repeat(MAX_CHAR + 1)}`,
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
      target: {
        value: `${"a".repeat(MAX_CHAR + 1)}`,
      },
    });

    fireEvent.click(screen.getByRole("checkbox"));

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(mockSubmit).not.toBeCalled();
    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe(
      "name"
    );
    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe(
      "surname"
    );
    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe(
      "test@mail.com"
    );
    expect(screen.getByPlaceholderText(/^password/i).value).toBe(
      `${"a".repeat(MAX_CHAR + 1)}`
    );
    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
      `${"a".repeat(MAX_CHAR + 1)}`
    );
    expect(screen.getByRole("checkbox").checked).toBe(true);
  });

  it("should display password not match error when password repeat does not match", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
      target: {
        value: "name",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
      target: {
        value: "surname",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
      target: {
        value: "passcode",
      },
    });

    fireEvent.click(screen.getByRole("checkbox"));

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(mockSubmit).not.toBeCalled();
    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe(
      "name"
    );
    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe(
      "surname"
    );
    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe(
      "test@mail.com"
    );
    expect(screen.getByPlaceholderText(/^password/i).value).toBe("password");
    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
      "passcode"
    );
    expect(screen.getByRole("checkbox").checked).toBe(true);
  });

  it("should display accept terms error when not accept terms", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
      target: {
        value: "name",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
      target: {
        value: "surname",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(mockSubmit).not.toBeCalled();
    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe(
      "name"
    );
    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe(
      "surname"
    );
    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe(
      "test@mail.com"
    );
    expect(screen.getByPlaceholderText(/^password/i).value).toBe("password");
    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
      "password"
    );
    expect(screen.getByRole("checkbox").checked).toBe(false);
  });

  it("should not display error when value is valid", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
      target: {
        value: "name",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
      target: {
        value: "surname",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
      target: {
        value: "password",
      },
    });

    fireEvent.click(screen.getByRole("checkbox"));

    fireEvent.submit(screen.getByRole("button"));

    await waitFor(() => expect(screen.queryAllByRole("alert")).toHaveLength(0));
    expect(mockSubmit).toBeCalledWith({
      firstName: "name",
      lastName: "surname",
      email: "test@mail.com",
      password: "password",
      confirmPassword: "password",
      acceptTerms: "accepted",
    });
    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe("");
    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe("");
    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe("");
    expect(screen.getByPlaceholderText(/^password/i).value).toBe("");
    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe("");
    expect(screen.getByRole("checkbox").checked).toBe(false);
  });
});
