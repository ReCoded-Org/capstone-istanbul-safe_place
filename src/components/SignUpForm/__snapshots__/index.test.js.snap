// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm renders correctly and matches the snapshot 1`] = `
<div
  className="signUpForm"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="formInputGroup"
    >
      <div
        className="formInput"
      >
        <input
          aria-label="firstName"
          name="firstName"
          placeholder="First name"
          type="text"
        />
      </div>
      <div
        className="formInput"
      >
        <input
          aria-label="lastName"
          name="lastName"
          placeholder="Last name"
          type="text"
        />
      </div>
    </div>
    <div
      className="formInput"
    >
      <input
        aria-label="email"
        name="email"
        placeholder="Your email"
        type="email"
      />
    </div>
    <div
      className="formInputGroup"
    >
      <div
        className="formInput"
      >
        <input
          aria-label="password"
          name="password"
          placeholder="Password"
          type="password"
        />
      </div>
      <div
        className="formInput"
      >
        <input
          aria-label="confirmPassword"
          name="confirmPassword"
          placeholder="Repeat password"
          type="password"
        />
      </div>
    </div>
    <div
      className="formInput"
    >
      <label
        htmlFor="acceptTerms"
      >
        <input
          id="acceptTerms"
          name="acceptTerms"
          type="checkbox"
          value="accepted"
        />
        I have read and agree with 
        <a
          href="/terms"
        >
          Terms of Service
        </a>
         and our 
        <a
          href="/policy"
        >
          Privecy Policy
        </a>
      </label>
    </div>
    <button
      className="submitBtn"
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
